draft-ietf-dnsop-svcb-https.md

```diff
@@ -1082,18 +1082,46 @@ single-client granularity.
 
 ## Interaction with Alt-Svc
 
-Clients that do not implement support for Encrypted ClientHello MAY
-skip the HTTPS RR query
-if a usable Alt-Svc value is available in the local cache.
-If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
-client connection procedure ({{client-behavior}}).
-
-Clients that implement support for ECH MUST perform the HTTPS RR query first,
-and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
-{{ech-client-behavior}}).
-
-This specification does not alter the DNS records used when connecting
-to an Alt-Svc hostname (typically A and/or AAAA only).
+Clients that implement support for both Alt-Svc and HTTPS records SHOULD
+retrieve any HTTPS records for the Alt-Svc alt-authority, and ensure that
+their connection attempts are consistent with both the Alt-Svc parameters
+and any received HTTPS SvcParams.  If present, the HTTPS record's target
+and port override the alt-authority.  For example, suppose that
+"https://example.com" sends an Alt-Svc field value of:
+
+~~~ HTTP
+Alt-Svc: h2="alt.example:443", h2="alt2.example:443", h3=":8443"
+~~~
+
+The client would retrieve the following HTTPS records:
+
+    alt.example.              IN HTTPS 1 . alpn=h2,h3 ech=...
+    alt2.example.             IN HTTPS 1 . alpn=h3 ech=...
+    _8443._https.example.com. IN HTTPS 1 alt3.example. (
+        port=9443 alpn=h2,h3 ech=... )
+
+The client could then attempt an HTTP/3 connection to `alt3.example:9443`
+with ECH, or an HTTP over TLS connection to `alt.example:443` with ECH,
+as these options are consistent with both an Alt-Svc field value and its
+HTTPS record.  It would not connect to `alt2.example`, as there is no
+possible connection that is consistent with both ALPN values ("h2" from
+Alt-Svc, "h3" from SvcParams), and the "ech" SvcParam triggers SVCB-reliant
+client behavior ({{ech-client-behavior}}).  Operators SHOULD avoid
```

Comment:
This reads like "the ech SvcParam triggers SVCB-reliant client behavior" is a reason for not choosing alt2.  But it can't be, because alt2 has a SVCB record.

Is there an error in the example?  Should alt2 not have the ech SvcParam, with the ech SvcParam on other records engaging the SVCB-reliant client behaviour?

Reply:
OK, adjusted.

Reply:
Oh, I completely missed the part where you connect based on the A/AAAA records.  Maybe again:

> an ECH-capable client would not attempt a direct h2 connection to "alt2.example" (that is, using the address listed in A/AAAA) because the "ech" SvcParam triggers SVCB-reliant client behavior ({{ech-client-behavior}})

If "direct" is a term of art, then the clarification might not be needed, but this is hardly "direct" by any reckoning familiar to me.

Reply:
OK, changed to

> An ECH-capable client would not connect to `alt2.example`,
> as there is no possible connection that is consistent with both ALPN values
> ("h2" from Alt-Svc, "h3" from SvcParams), and the "ech" SvcParam prevents
> the client from falling back to a plain A/AAAA-based connection (i.e. the
> client is SVCB-reliant, {{ech-client-behavior}}).

Please review.

Reply:
I'm fine with "direct" or "plain A/AAAA".  Either seems clear enough to me, but maybe my understanding has just been tainted by informal usage of "direct" for this meaning in these various discussions.  

Related note: If we don't like "direct" to refer to plain A/AAAA, might also want to reword Section 9.1: 

> The SVCB-optional client behavior specified in Section 3 permits clients to fall back to a direct connection if all SVCB options fail.

Reply:
I've restructured this section to avoid the need for this terminology.  Please review.